### ISSUE TITLE
Cannot use special characters in user password: "UnicodeEncodeError: 'ascii' codec can't encode character u'\xea' in position 19: ordinal not in range(128)"

### ISSUE THREAD

**Issue by Gwend4l** [matrix-org/synapse#1814]
- OS: Debian Jessie
- Synapse version: 0.18.7-1
- Python2 version: 2.7.9
- Python3 version: 3.4.2

On my synapse server, a user first registered with a very simple password based on regular ASCII characters. She could log in and use Riot without any issue.

Later, this user changer her password to something much longer, _and with special characters, such as 'ê'._

The password change worked. However, after doing that, she could not log in anymore with her account (with the new password), and faced an error "Internal Error 500" (or something like that) on Riot on all her devices.

Looking at the logs, I discovered that the problem seemed to be on Synapse's side: each time she tried to log in with her new password containing special chars, this happened in the logs:

~~~~
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/synapse/http/server.py", line 116, in wrapped_request_handler
	yield request_handler(self, request, request_metrics)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
	result = result.throwExceptionIntoGenerator(g)
  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	return g.throw(self.type, self.value, self.tb)
  File "/usr/lib/python2.7/dist-packages/synapse/http/server.py", line 255, in _async_render
	callback_return = yield callback(request, **kwargs)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
	result = result.throwExceptionIntoGenerator(g)
  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	return g.throw(self.type, self.value, self.tb)
  File "/usr/lib/python2.7/dist-packages/synapse/rest/client/v1/login.py", line 94, in on_POST
	result = yield self.do_password_login(login_submission)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
	result = result.throwExceptionIntoGenerator(g)
  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	return g.throw(self.type, self.value, self.tb)
  File "/usr/lib/python2.7/dist-packages/synapse/rest/client/v1/login.py", line 137, in do_password_login
	password=login_submission["password"],
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
	result = result.throwExceptionIntoGenerator(g)
  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	return g.throw(self.type, self.value, self.tb)
  File "/usr/lib/python2.7/dist-packages/synapse/handlers/auth.py", line 495, in _check_password
	canonical_user_id = yield self._check_local_password(user_id, password)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1128, in _inlineCallbacks
	result = g.send(result)
  File "/usr/lib/python2.7/dist-packages/synapse/handlers/auth.py", line 524, in _check_local_password
	result = self.validate_hash(password, password_hash)
  File "/usr/lib/python2.7/dist-packages/synapse/handlers/auth.py", line 660, in validate_hash
	stored_hash.encode('utf-8')) == stored_hash
UnicodeEncodeError: 'ascii' codec can't encode character u'\xea' in position 19: ordinal not in range(128)
~~~~

So it would seem that logging in with a strange password makes Synapse crash?

**Comment by Gwend4l** [comment]
As a workaround, we managed to reset the password using the "Forgot your password?" link on Riot, and remove the special character from the password.

**Comment by richvdh** [comment]
should be fixed by https://github.com/matrix-org/synapse/pull/1711, once that lands

**Comment by Gwend4l** [comment]
Oh sorry, I had not found the duplicate when posting the bug. But great news, thanks to all for the fix.

**Comment by richvdh** [comment]
Hope this was fixed by #1711
